ghcide/src/Development/IDE/Plugin/SelectionRange.hs

```diff
@@ -0,0 +1,76 @@
+import qualified Data.Map.Strict                 as Map
+import           Data.Maybe                      (fromMaybe, mapMaybe)
+import           Development.IDE                 (GetHieAst (GetHieAst),
+                                                  HieAstResult (HAR, hieAst),
+                                                  IdeAction,
+                                                  IdeState (shakeExtras),
+                                                  Range (Range),
+                                                  fromNormalizedFilePath,
+                                                  realSrcSpanToRange,
+                                                  runIdeAction,
+                                                  toNormalizedFilePath',
+                                                  uriToFilePath')
+import           Development.IDE.Core.Actions    (useE)
+import           Development.IDE.GHC.Compat      (HieAST (Node), Span, getAsts)
+import           Development.IDE.GHC.Compat.Util (mkFastString)
+import           Ide.Types                       (PluginDescriptor (pluginHandlers),
+                                                  PluginId,
+                                                  defaultPluginDescriptor,
+                                                  mkPluginHandler)
+import           Language.LSP.Server             (LspM)
+import           Language.LSP.Types              (List (List),
+                                                  NormalizedFilePath, Position,
+                                                  ResponseError,
+                                                  SMethod (STextDocumentSelectionRange),
+                                                  SelectionRange (..),
+                                                  SelectionRangeParams (..),
+                                                  TextDocumentIdentifier (TextDocumentIdentifier))
+
+descriptor :: PluginId -> PluginDescriptor IdeState
+descriptor plId = (defaultPluginDescriptor plId)
+    { pluginHandlers = mkPluginHandler STextDocumentSelectionRange selectionRangeHandler
+    }
+
+selectionRangeHandler :: IdeState -> PluginId -> SelectionRangeParams -> LspM c (Either ResponseError (List SelectionRange))
+selectionRangeHandler ide _ SelectionRangeParams{..} = do
+    let (TextDocumentIdentifier uri) = _textDocument
+    let filePathMaybe = toNormalizedFilePath' <$> uriToFilePath' uri
+    case filePathMaybe of
+        Nothing -> pure . Right . List $ []
+        Just filePath -> liftIO $ do
+            let (List positions) = _positions
+            selectionRanges <- runIdeAction "SelectionRange" (shakeExtras ide) $ getSelectionRanges filePath positions
+            pure . Right . List $ selectionRanges
+
+getSelectionRanges :: NormalizedFilePath -> [Position] -> IdeAction [SelectionRange]
+getSelectionRanges file positions = fmap (fromMaybe []) <$> runMaybeT $ do
+    (HAR{hieAst}, _) <- useE GetHieAst file
```

Comment:
You shouldn't ignore the `PositionMapping` and instead use `oldPositionToNew` to adjust all the output positions and `newPositionToOld` to adjust all the input positions.

Reply:
Thanks for pointing this out! I will fix this later.

Reply:
There seems to be no such function as `oldPositionToNew` through. Having refered to other plugins, I use 
 functions like `toCurrentPosition` to apply position mapping to input and output. Resolving this conversation for now.